File: proteus/patch/Proteus1Map.swift

extension Proteus1 {
  
  enum Map {
    
    static let patchTruss = try! SinglePatchTruss("proteus1.map", 256, params: parms.params(), initFile: "proteus1-map-init", createFileData: {
      sysexData($0, deviceId: 0)
    }, parseOffset: 5, pack: { bodyData, param, value in
      Proteus.pack(&bodyData, parm: param.p! - 512, value: value)
    }, unpack: { bodyData, param in
      Proteus.unpack(bodyData, parm: param.p! - 512)
    })
    
    static let patchTransform: MidiTransform = .single(throttle: 100, Proteus.deviceId, .patch( param: { editorVal, bodyData, parm, value in
      return [(.sysex(Proteus.paramData(parm: parm.p!, value: value)), 10)]
    }, patch: { editorVal, bodyData in
      [(.sysex(sysexData(bodyData, deviceId: UInt8(editorVal))), 10)]
    }, name: nil))

    static func sysexData(_ bytes: [UInt8], deviceId: UInt8) -> [UInt8] {
      Proteus.sysex(deviceId: deviceId, [0x07] + bytes)
    }

    static let params = parms.params()
    
    static let parms: [Parm] = .prefix([], count: 128, bx: 1) { _ in
      [.p([], p: 512, .max(191))]
    }
  }
  
}

File: proteus/proteus1_map.js

const sysexData = Proteus.sysex([0x07, 'b'])

const patchTransform = {
  type: 'singlePatch',
  throttle: 100, 
  param: (path, parm, value) => [[Proteus.paramData(parm.p, value), 10]], 
  patch: [[sysexData, 10]],
}

const parms = [
  { prefix: '', count: 128, bx: 1, block: [
    ['', { p: 512, max: 191 }],
  ] },
]

const patchTruss = {
  single: "proteus1.map", 
  bodyDataCount: 256, 
  parms: parms, 
  initFile: "proteus1-map-init", 
  createFileData: sysexData, 
  parseBody: 5, 
  pack: (parm, value) => Proteus.pack(parm.p - 512, value),
  unpack: (parm) => Proteus.unpack(parm.p - 512),
}

module.exports = {
  patchTruss,
  patchTransform,
}
